### ISSUE TITLE
java.lang.IllegalArgumentException: Invalid scanline stride with a high resolution PSB file

### ISSUE THREAD

**Issue by vsarya** [haraldk/TwelveMonkeys#398]
Following exception is observed with this high resolution PSB [file](https://drive.google.com/open?id=1-TU3HAf5WjF-Z8ZDLWE9TMDYDdlR3ieJ).

java.lang.IllegalArgumentException: Invalid scanline stride
	at java.awt.image.ComponentSampleModel.getBufferSize(ComponentSampleModel.java:274)
	at java.awt.image.ComponentSampleModel.verify(ComponentSampleModel.java:252)
	at java.awt.image.ComponentSampleModel.<init>(ComponentSampleModel.java:170)
	at java.awt.image.PixelInterleavedSampleModel.<init>(PixelInterleavedSampleModel.java:87)
	at java.awt.image.PixelInterleavedSampleModel.createCompatibleSampleModel(PixelInterleavedSampleModel.java:144)
	at javax.imageio.ImageTypeSpecifier.getSampleModel(ImageTypeSpecifier.java:1039)
	at javax.imageio.ImageTypeSpecifier.createBufferedImage(ImageTypeSpecifier.java:1072)
	at com.twelvemonkeys.imageio.ImageReaderBase.getDestination(Unknown Source)
	at com.twelvemonkeys.imageio.plugins.psd.PSDImageReader.read(PSDImageReader.java:357)

**Comment by haraldk** [comment]
Hi Vishal,

I'm afraid the PSD/PSB in question is too large to open as a `BufferedImage`. Unfortunately, this can't be fixed easily, as `BufferedImage`s are backed by arrays, and arrays in Java have a size limit. This is not a bug. Though I agree, the exception message should be better, which I consider a bug.

That said, there are options to read this image using Java, the ImageIO API and our PSD plugin. :-)

* If you need the full resolution, you can read the image in smaller pieces, using the `ImageReadParam.setSourceRegion(...)` method. This allows you to read the entire image into multiple `BufferedImage`s. If you have enough RAM, you could create a tiled `RenderedImage` containing the entire image data. 

* If you don't need the full resolution (this image is huuuuge),  you can read the image sub sampled, using `ImageReadParam.setSourceSubSampling(...)`. This allows you to read a "down scaled" version of the image into a single `BufferedImage`.  This is probably more manageable. 


Best regards,

-- 
Harald K

**Comment by dmichulke** [comment]
Related: https://bugs.openjdk.java.net/browse/JDK-8078589

TLDR: 30K * 30K * 3 > Integer.MAX_INT (InterleavedSampleModel uses at least 3 bytes for a pixel)

**Comment by haraldk** [comment]
@dmichulke Thanks,

(I incorrectly closed this some days ago...)

The only thing I plan to do, is giving a better exception message (that is the bug). 
Decoding the linked image to a `BufferedImage` with `InterleavedSampleModel` is impossible. 


Best regards,

-- 
Harald K

**Comment by DavidPDP** [comment]
I've the same problem with this image: https://www.spacetelescope.org/images/heic1502a/ 
Its dimensions are: 69536 * 22230 px, and its format is PSB.
This is the line of error code 274 and I don't see the error depends on computer's RAM because I'm running the code with 30 GB of RAM:
```java
if (scanlineStride < 0 || scanlineStride > (Integer.MAX_VALUE / height)){
   throw new IllegalArgumentException("Invalid scanline stride")
} 
```
```java
-Xmx30720m
```
Also checking the buffer creation code, the limitation should be width * height < Integer.MAX_VALUE. So What could be done to read the image?
Thanks.

**Comment by haraldk** [comment]
@DavidPDP 

You will have this problem with any image where image dimensions multiplied with the number of samples per pixel is larger than `Integer.MAX_VALUE`. The test you found in the code is a little non-intuitive to avoid integer overflow.

I've listed three options above, feel free to try those. As long as you can create a new `BufferedImage` with the dimensions you want and compatible pixel layout, you can pass that as a destination in the `ImageReadParam` you should be good. But no standard `BufferedImage` type will work though. I recommend reading in smaller parts, or subsampled.


Best regards,

-- 
Harald K

**Comment by DavidPDP** [comment]
@haraldk
Excuse me, I've been trying to load the image with this code:
```
param.setDestinationType(ImageTypeSpecifier.createBanded(ColorSpace.getInstance(ColorSpace.CS_sRGB), 
                                new int[]{0, 1, 2}, new int[] {0, 0, 0}, DataBuffer.TYPE_BYTE, false, false));
```
But it throws me the following error:
```
javax.imageio.IIOException: Destination type from ImageReadParam does not match legal imageTypes from reader: javax.imageio.ImageTypeSpecifier$Banded@2a2a1d44
	at com.twelvemonkeys.imageio.ImageReaderBase.getDestination(ImageReaderBase.java:302)
	at com.twelvemonkeys.imageio.plugins.psd.PSDImageReader.read(PSDImageReader.java:370)
```
if I change it to just RGB type (ColorSpace):
```
param.setDestinationType(ImageTypeSpecifier.createBanded(ColorSpace.getInstance(ColorSpace.TYPE_RGB), 
	                    new int[]{0, 1, 2}, new int[] {0, 0, 0}, DataBuffer.TYPE_BYTE, false, false));
```
It throws me:
```
Exception in thread "main" java.lang.IllegalArgumentException: Unknown color space
	at java.awt.color.ColorSpace.getInstance(ColorSpace.java:360)
```
maybe, are you know what I can do to load the image in arrays with its different color channels?

**Comment by haraldk** [comment]
@DavidPDP The above code works for me (the first version, the second can't work obviously, you have to use one of the `CS_*` constants when creating a `ColorSpace`). I used Java 8 (1.8.0_x).

But I tested again, using a different Java version (OpenJDK 11.0.1+x). There seems to be a JDK issue, where something has changed in the comparison of `ImageTypeSpecifier`s, so that two equal (but not the same) specifiers are no longer considered equal... That's unfortunate. 

For now, you can still get the same effect, by using the following code:

````java
Iterator<ImageTypeSpecifier> types = reader.getImageTypes(imageNo);
while (types.hasNext()) {
    ImageTypeSpecifier candidate = types.next();
    if (candidate.getSampleModel() instanceof BandedSampleModel) {
        param.setDestinationType(candidate);
        break;
    }
}
````

This works for me, both Java 8 and 11.

-- 
Harald K

**Comment by DavidPDP** [comment]
@haraldk 
Hi, I understand and the last code worked for me, with the iterator. The first one, I executed it again in the following environment and it didn't work.
```
openjdk version "1.8.0_191"
OpenJDK Runtime Environment (build 1.8.0_191-8u191-b12-2ubuntu0.18.10.1-b12)
OpenJDK 64-Bit Server VM (build 25.191-b12, mixed mode)
```
Also, I want to ask you how the code works? As far as I understand, the parameter helps me to be able to divide into arrays each of image's color channels, but how should I read it?
```
try {
			ImageInputStream input = ImageIO.createImageInputStream(new File(path));
			Iterator<ImageReader> readers = ImageIO.getImageReaders(input);
			if (!readers.hasNext()) {
		        throw new Exception();
		    }else {	
	    		reader = readers.next();
	    		reader.setInput(input);
	    		
	    		Iterator<ImageTypeSpecifier> types = reader.getImageTypes(0);
	    		while (types.hasNext()) {
	    		    ImageTypeSpecifier candidate = types.next();
	    		    if (candidate.getSampleModel() instanceof BandedSampleModel) {
	    		        param.setDestinationType(candidate);
	    		        break;
	    		    }
	    		}

	    		nasaImage = reader.read(0,param);
		    }
		} catch (IOException e) {
			e.printStackTrace();
		}
```
I ask you because I don't understand how the image should be read with the parameter and how could I access each array to retrieve the colors later?
Thanks.

**Comment by haraldk** [comment]
@DavidPDP 

Yes, sorry, OpenJDK 8 uses the same color management as Oracle Java 9+, so you'll have the same issue there. I updated the comment to make it clear I used Oracle Java 8. I also found a simpler workaround and added a quick explanation.

If you have more questions about similar issues, and to prevent further deviation from the original issue, I kindly ask you to either 1) ask a question describing the problem on StackOverflow and tag it with `javax.imageio`, `bufferedimage` and/or `twelvemonkeys`  (as appropriate) *OR* 2) if you have found a bug or missing feature related to the TwelveMonkeys library, file a new issue here.


Best regards,

-- 
Harald K
